### PR TITLE
Add proper integration for AsyncHttpClient 2+

### PR DESCRIPTION
**Pull request by iNikem**
Based on and blocked by #2323

Only `instrumentation/async-http-client-2.0` part is new. In addition Play high concurrency test is now fixed.

### CODE REVIEW

File: instrumentation/play/play-2.4/javaagent/src/test/groovy/client/PlayWsClientTest.groovy [hunk 1]
```diff
@@ -58,8 +59,4 @@ class PlayWsClientTest extends HttpClientTest implements AgentTestTrait {
     return false
   }
 
-  @Override
-  boolean testCausality() {
-    return false
-  }
```
Comment (trask) [inline instrumentation/play/play-2.4/javaagent/src/test/groovy/client/PlayWsClientTest.groovy]: 💯

File: instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/ResponseAdvice.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+import org.asynchttpclient.Response;
+
+public class ResponseAdvice {
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static Scope onEnter(
+      @Advice.This AsyncCompletionHandler<?> handler, @Advice.Argument(0) Response response) {
+
+    // TODO I think all this should happen on exit, not on enter.
+    // After response was handled by user provided handler.
```
Comment (trask) [inline instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/ResponseAdvice.java:25]: I thought you only thought this for sync handlers?

**Review by trask — APPROVED** [review]

File: instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/RequestAdvice.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+      @Advice.Argument(1) AsyncHandler<?> handler,
+      @Advice.Local("otelScope") Scope scope) {
+    Context parentContext = currentContext();
+    if (!tracer().shouldStartSpan(parentContext)) {
+      return;
+    }
+
+    Context context = tracer().startSpan(parentContext, request, request);
+    InstrumentationContext.get(AsyncHandler.class, Pair.class)
+        .put(handler, Pair.of(parentContext, context));
```
[the remark this answers is not in the document]
Comment (iNikem) [reply instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/RequestAdvice.java:33]: I will take a look, but this is just a copy-paste from AHC 1.9 integration

File: instrumentation/async-http-client-2.0/javaagent/async-http-client-2.0-javaagent.gradle [hunk 4]
```diff
@@ -0,0 +1,20 @@
+ext {
+  maxJavaVersionForTests = JavaVersion.VERSION_1_8
```
[the remark this answers is not in the document]
Comment (iNikem) [reply instrumentation/async-http-client-2.0/javaagent/async-http-client-2.0-javaagent.gradle:2]: Added

File: instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/ResponseAdvice.java (hunk 2 above)
Comment (iNikem) [reply instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/ResponseAdvice.java:25]: Removed this comment

Comment (iNikem) [reply instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/ResponseAdvice.java:25]: We talked about this recently in some other PR. About in what context response processing should take place. You convinced me :) I will remove this comment from AHC 1.9 as well.

File: instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/RequestAdvice.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (iNikem) [reply instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/RequestAdvice.java:33]: I am not 100% sure which method is better. Wrapping seems cleaner and less bytecode magic. But we already have some problems with changing runtime types. And in this case `ResponseInstrumentation` actually addresses just some subclasses of `AsyncHandler` so I am not sure... 

Will not do that in this PR

[the remark this answers is not in the document]
Comment (trask) [reply instrumentation/async-http-client-2.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/asynchttpclient/RequestAdvice.java:33]: makes sense

**Pull request merged**